doc/syslog/syslog-rate-limit-design.md

```diff
@@ -0,0 +1,383 @@
+
+This feature allows user to configure SystemLogRateLimitInterval and SystemLogRateLimitBurst for host, containers.
+
+### Requirements
+
+- Support syslog message rate limit configuration for host and containers
+- New CLI for rate limit configuration
+- Have rate limit configuration persistent
+- Default rate limit shall be applied if no configuration provided
+- Both multi ASIC and single ASIC platform shall use rsyslog-container.conf.j2 as template
+- APP extension shall be able to declare its capability of this feature
+- APP extension is responsible for reading/listening rate limit configuration from CONFIG DB and set rsyslog configuration accordingly.
+
+### Architecture Design
+
+1. CLI is only responsible for putting rsyslog configuration to CONFIG DB
+2. hostcfgd shall be extended to handle host side rsyslog configuration by listening CONFIG DB change
+3. A new daemon containercfgd shall be added to each container to handle container side rsyslog configuration
+4. rsyslog.conf.j2 and rsyslog-container.conf.j2 shall be extended to accept template variable for SystemLogRateLimitInterval and SystemLogRateLimitBurst.
+5. rsyslog.conf for single ASIC platform will be removed and replaced by rsyslog-container.conf.j2
+6. App extension shall handle rsyslog configuration itself
+
+#### CLI flow
+
+![cli_rate_limit_config_flow](/doc/syslog/images/cli_rate_limit_config_flow.svg)
+
+#### Host side flow
+
+##### Loading config from CONFIG DB
+
+There is an existing service rsyslog-config.service. On switch start up, this service renders rsyslog.conf.j2 and restarts rsyslog service. No extra changes for this feature.
+
+![host_rate_limit_load_flow](/doc/syslog/images/host_rate_limit_load_flow.svg)
+
+##### Runtime configuration
+
+![host_config_on_fly](/doc/syslog/images/host_rate_limit_config_flow.svg).
+
+#### Container side flow
+
+##### Loading config from CONFIG DB
+
+In SONiC, container start up script is generated by rendering docker_image_ctl.j2. In each container start up script, there are 3 phases: `preStartAction`, `start container`, `postStartAction`. In phase `preStartAction`, there is an existing function `updateSyslogConf`. `updateSyslogConf` shall be extended to render rsyslog-container.conf.j2 for both single ASIC and multi ASIC platforms. Flow for `updateSyslogConf`:
+
+![container_rate_limit_load_flow](/doc/syslog/images/container_rate_limit_load_flow.svg).
+
+> Note: docker support copying file to a stopped container
+
+##### Runtime configuration
+
+![container_config_on_fly](/doc/syslog/images/container_rate_limit_config_flow.svg).
+
+> Note: according to test, syslog rate limit configuration on host side would not affect container side.
```

Comment:
@Junchao-Mellanox IMHO, it's bit confusing comparing to how it's done in AUTO TECHSUPPORT, where global configuration overrides feature rate limit configuration. Here, using 'GLOBAL', does not reflect the actual intentions.

Reply:
How about HOST?

Reply:
@Junchao-Mellanox unfortunately i don't have neither a strong opinion here, nor a good suggestion